### DISCUSSION TITLE
Reverse a rotational model

### DISCUSSION THREAD

**Discussion by s-celles** [OpenModelica/OpenModelica#11192] [Q&A] [answered]
Hello,

I'm new to Modelica. I'd like to answer the following problem using Modelica ie by reversing a rotational model

Here is description of the problem:

Motor speed of a CNC machine is 1500 rpm.

When the emergency stop is triggered, the machine is immediately powered off, but it takes 15 seconds for the rotation of the mechanical part of the machine to stop.

The motor rotation speed profile (in tr/min ie rpm) of a CNC machine is shown below.
![image](https://github.com/OpenModelica/OpenModelica/assets/109167/c4f93de9-bf82-4b7d-bc17-13dd98ada4f6)

During this operation, the resistive torque (Tr = 0.7 N.m) on the shaft slows down the motor
When an emergency stop is triggered, the useful torque of the motor driving the machine tool is zero.

1. Calculate the acceleration dΩ/dt of the motor in rad.s^-2.
2. Deduce the value of the moment of inertia of the motor.

By hand I answered the question and found

1. dΩ/dt = -10.472 rad.s^-2   or dn/dt = -100 tr.min^-2
2. J dΩ/dt = -Tr so J = -0.7/-10.472 = 0.0668 kg.m^2

So I can model the problem like so

![image](https://github.com/OpenModelica/OpenModelica/assets/109167/e157204c-d8df-4d2a-a8d3-e9a3aea46fb7)

```modelica
model freinage_machine
  Modelica.Mechanics.Rotational.Sensors.SpeedSensor speedSensor annotation(
    Placement(visible = true, transformation(origin = {48, 24}, extent = {{-10, -10}, {10, 10}}, rotation = 0)));
  Modelica.Mechanics.Rotational.Sources.ConstantTorque constantTorque(tau_constant = -0.7)  annotation(
    Placement(visible = true, transformation(origin = {-26, 24}, extent = {{-10, -10}, {10, 10}}, rotation = 0)));
  Modelica.Mechanics.Rotational.Components.Inertia inertia(J = 0.06684491978, w(displayUnit = "rpm", start = 157.0796326794897))  annotation(
    Placement(visible = true, transformation(origin = {8, 24}, extent = {{-10, -10}, {10, 10}}, rotation = 0)));
equation
  connect(constantTorque.flange, inertia.flange_a) annotation(
    Line(points = {{-16, 24}, {-2, 24}}));
  connect(inertia.flange_b, speedSensor.flange) annotation(
    Line(points = {{18, 24}, {38, 24}}));
  annotation(
    uses(Modelica(version = "4.0.0")));
end freinage_machine;
```

![image](https://github.com/OpenModelica/OpenModelica/assets/109167/447ee88c-b32c-4070-863b-69562cea37a7)

rotationnal speed is 0 rpm at 15 s which is correct.

I wonder if there is a way to "reverse" model to answer the questions.

So I started with a `CombiTimeTable` for rotational speed (in rpm)
![image](https://github.com/OpenModelica/OpenModelica/assets/109167/b3987c03-cd61-46ff-8a35-a76d4624a57c)
```modelica
model temp
  Modelica.Mechanics.Rotational.Sources.Speed speed annotation(
    Placement(visible = true, transformation(origin = {-20, -20}, extent = {{-10, -10}, {10, 10}}, rotation = 0)));
  Modelica.Blocks.Sources.CombiTimeTable combiTimeTable(table = [0, 1500; 15; 0]) annotation(
    Placement(visible = true, transformation(origin = {-56, -20}, extent = {{-10, -10}, {10, 10}}, rotation = 0)));
equation
  connect(combiTimeTable.y[1], speed.w_ref) annotation(
    Line(points = {{-45, -20}, {-33, -20}}, color = {0, 0, 127}));
  annotation(
    uses(Modelica(version = "4.0.0")));end temp;
```

But I can't know how to complete such a "reverse" model ie how to get inertia J = 0.0668 kg.m^2

Ideally I would like this reverse model "feed" the normal model with inertia as a variable/parameter.

Any idea?

Sorry for my newbie question.

Kind regards

**Comment by s-celles** [comment]
I tried this...

![image](https://github.com/OpenModelica/OpenModelica/assets/109167/a5e4fa83-6869-4dca-8992-ed3dacd5bd18)

```
model freinage_machine
  Modelica.Mechanics.Rotational.Sensors.SpeedSensor speedSensor annotation(
    Placement(visible = true, transformation(origin = {48, 24}, extent = {{-10, -10}, {10, 10}}, rotation = 0)));
  Modelica.Mechanics.Rotational.Sources.ConstantTorque constantTorque(tau_constant = -0.7)  annotation(
    Placement(visible = true, transformation(origin = {-26, 24}, extent = {{-10, -10}, {10, 10}}, rotation = 0)));
  Modelica.Mechanics.Rotational.Components.Inertia inertia(J = 0.06684491978, w(displayUnit = "rpm", start = 157.0796326794897))  annotation(
    Placement(visible = true, transformation(origin = {8, 24}, extent = {{-10, -10}, {10, 10}}, rotation = 0)));
  Modelica.Blocks.Sources.CombiTimeTable combiTimeTable(table = [0, 1500; 10, 1500; 25, 0; 30, 0])  annotation(
    Placement(visible = true, transformation(origin = {6, -42}, extent = {{-10, -10}, {10, 10}}, rotation = 0)));
  Modelica.Blocks.Continuous.Derivative derivative annotation(
    Placement(visible = true, transformation(origin = {88, -42}, extent = {{-10, -10}, {10, 10}}, rotation = 0)));
  Modelica.Blocks.Sources.RealExpression realExpression(y = -0.7)  annotation(
    Placement(visible = true, transformation(origin = {30, -8}, extent = {{-10, -10}, {10, 10}}, rotation = 0)));
  Modelica.Blocks.Math.Division division annotation(
    Placement(visible = true, transformation(origin = {136, -24}, extent = {{-10, -10}, {10, 10}}, rotation = 0)));
  Modelica.Blocks.Math.MultiProduct multiProduct(nu = 2)  annotation(
    Placement(visible = true, transformation(origin = {50, -42}, extent = {{-6, -6}, {6, 6}}, rotation = 0)));
  Modelica.Blocks.Sources.RealExpression realExpression1(y = 2*Modelica.Constants.pi/60) annotation(
    Placement(visible = true, transformation(origin = {-67, -69}, extent = {{-83, -7}, {83, 7}}, rotation = 0)));
equation
  connect(constantTorque.flange, inertia.flange_a) annotation(
    Line(points = {{-16, 24}, {-2, 24}}));
  connect(inertia.flange_b, speedSensor.flange) annotation(
    Line(points = {{18, 24}, {38, 24}}));
  connect(realExpression.y, division.u1) annotation(
    Line(points = {{41, -8}, {95, -8}, {95, -18}, {123, -18}}, color = {0, 0, 127}));
  connect(derivative.y, division.u2) annotation(
    Line(points = {{99, -42}, {109, -42}, {109, -30}, {123, -30}}, color = {0, 0, 127}));
  connect(multiProduct.u[1], combiTimeTable.y[1]) annotation(
    Line(points = {{44, -42}, {18, -42}}, color = {0, 0, 127}));
  connect(multiProduct.y, derivative.u) annotation(
    Line(points = {{57.02, -42}, {76.02, -42}}, color = {0, 0, 127}));
  connect(realExpression1.y, multiProduct.u[2]) annotation(
    Line(points = {{24.3, -69}, {30.3, -69}, {30.3, -42}, {44.3, -42}}, color = {0, 0, 127}));
  annotation(
    uses(Modelica(version = "4.0.0")),
  Diagram(coordinateSystem(extent = {{-200, -100}, {200, 100}})),
  Icon(coordinateSystem(extent = {{-200, -100}, {200, 100}})),
  version = "");
end freinage_machine;
```

and found the value of inertia between 10s and 25s on division.y output ie 0.0668 kg.m^2.

But I don't think that's a proper way of reversing the model.

Any suggesting will be very nice.

**Comment by casella** [comment, marked as answer]
As I understand, reversing models in the sense that you swap time-varying inputs and outputs is something that all Modelica tools do. You only need to make sure that the inverse problem has a solution (e.g. it shouldn't contain Dirac impulses).

Computing the inertia is something different. The inertia is a fixed parameter, not a time-varying variable. It can't change during the simulation. You could set up an _optimization_ problem where the inertia J is a parameter to be identified, and use e.g. least-squares algorithms to do it. There are Modelica tools that do that, but it's not a simulation problem.

Does this answer your question?

**Comment by s-celles** [comment]
This is very clear.

But I wonder if anwer can't be done using OpenModelica Optimization Editor which seems to be unmaintained according https://github.com/OpenModelica/OpenModelica/discussions/11025

That's sad.

**Reply by casella** [reply]
Sorry about that...
